src/Lecture3.hs

```diff
@@ -137,6 +156,13 @@ data List1 a = List1 a [a]
 
 -- | This should be list append.
 instance Semigroup (List1 a) where
+  (List1 a [b]) <> (List1 c [d]) = List1 a ([b] ++ [c] ++ [d])
+  (List1 a [b]) <> (List1 c []) = List1 a (b : [c])
+  (List1 a []) <> (List1 b (x:xs)) = List1 a (b:x:xs)
+  (List1 a []) <> (List1 c []) = List1 a [c]
+  (List1 a (x:xs)) <> (List1 b (y1: y2: ys)) = List1 a ((x:xs) ++ [b] ++ (y1:y2:ys))
+  (List1 a (x1:x2:xs)) <> (List1 b (y:ys)) = List1 a ((x1:x2:xs) ++ [b] ++ (y:ys))
+  (List1 a (x1:x2:xs)) <> (List1 b []) = List1 a ((x1:x2:xs) ++ [b])
```

Comment:
I see that you handle lots of redundant patterns. In all patterns you put `a` as the first element of the final list and concatenate all the remaining lists. This can be written as a single pattern.

Consider starting with the following:

```haskell
  (List1 a xs) <> (List1 b ys) = List1 a ...
```

And then do the rest from this single pattern.